### PR TITLE
[Concurrency] Add a builtin to get the current task in an async function

### PR DESCRIPTION
**Pull request by DougGregor**
This introduces a new builtin, `getCurrentAsyncTask()`, that produces a
reference to the current task. This builtin can only be used within
`async` functions, and IR generation merely grabs the task argument
and packages it up.

The type of this function is `() -> Builtin.NativeObject`, because we
don't currently have a Swift-level representation of tasks, and can
probably handle everything through builtins or runtime calls.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/SIL/IR/OperandOwnership.cpp [hunk 1]
```diff
@@ -1033,6 +1033,7 @@ ANY_OWNERSHIP_BUILTIN(IntInstrprofIncrement)
   }
 CONSTANT_OWNERSHIP_BUILTIN(Owned, MustBeInvalidated, COWBufferForReading)
 CONSTANT_OWNERSHIP_BUILTIN(Owned, MustBeInvalidated, UnsafeGuaranteed)
+CONSTANT_OWNERSHIP_BUILTIN(Owned, MustBeInvalidated, GetCurrentAsyncTask)
```
Comment (gottesmm) [inline lib/SIL/IR/OperandOwnership.cpp:1036]: This isn't needed. This is only needed to describe the ownership of arguments. No arguments, this isn't needed.

File: lib/SILGen/SILGenFunction.h [hunk 2]
```diff
@@ -2000,7 +2000,10 @@ class LLVM_LIBRARY_VISIBILITY SILGenFunction
   
   /// Enter a cleanup to emit a ReleaseValue/DestroyAddr of the specified value.
   CleanupHandle enterDestroyCleanup(SILValue valueOrAddr);
-  
+
+  /// Enter cleanup to emit an EndLifetime operation for the given value.
```
Comment (gottesmm) [inline lib/SILGen/SILGenFunction.h:2004]: A few thoughts:

1. Can you hide this and create an API that just returns a ManagedValue. We don't want people to create this by hand.

```
/// Return an owned managed value for \p value that is cleaned up using an end_lifetime instruction.
///
/// The end_lifetime cleanup is not placed into the ManagedValue itself and thus can not be forwarded.
/// This means that the ManagedValue is treated as a +0 value. This means that the owned value will be copied
/// by SILGen if it is ever needed as a +1 value (meaning any time that the value escapes).
///
/// DISCUSSION: end_lifetime ends the lifetime of an owned value in OSSA without resulting in a destroy
/// being emitted. This cleanup should only be used for owned values that do not need to be destroyed
/// if they do not escape the current call frame but need to be copied if they escape.
ManagedValue emitManagedRValueWithEndLifetimeCleanup(SILValue value);
```

File: lib/SILGen/SILGenDecl.cpp [hunk 3]
```diff
@@ -1326,6 +1326,32 @@ CleanupHandle SILGenFunction::enterDestroyCleanup(SILValue valueOrAddr) {
...
+    llvm::errs() << "EndLifetimeCleanup\n"
+                 << "State:" << getState() << "\n"
+                 << "Value:" << v << "\n";
+#endif
+  }
+};
+} // end anonymous namespace
+
+CleanupHandle SILGenFunction::enterEndLifetimeCleanup(SILValue value) {
+  Cleanups.pushCleanup<EndLifetimeCleanup>(value);
```
Comment (gottesmm) [inline lib/SILGen/SILGenDecl.cpp:1351]: See comment below. I would just have something that emits a ManagedValue directly and the managed value.

File: lib/SILGen/SILGenFunction.h (hunk 2 above)
Comment (DougGregor) [reply lib/SILGen/SILGenFunction.h:2004]: Okay!

File: lib/SIL/IR/OperandOwnership.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/SIL/IR/OperandOwnership.cpp]: Sure, updated with a "should not be called".

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Pull request merged**